src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceUpdateService.java

```diff
@@ -0,0 +1,206 @@
+                } catch (Exception e) {
+                    log.error("Failed to delete an index [{}]", index, e);
+                }
+            }
+            if (!deletedIndices.isEmpty()) {
+                parameter.getIndices().removeAll(deletedIndices);
+                datasourceFacade.updateDatasource(parameter);
+            }
+        } catch (Exception e) {
+            log.error("Failed to delete old indices for {}", parameter.getId(), e);
+        }
+    }
+
+    /**
+     * Validate the next csv record as header
+     * This method move iterator one step forward
+     *
+     * @param iterator the csv record iterator
+     * @return next CSVRecord
+     */
+    private CSVRecord validateAndReturnHeader(Iterator<CSVRecord> iterator) {
+        if (iterator.hasNext() == false) {
+            throw new OpenSearchException("geoip database is empty");
+        }
+        CSVRecord header = iterator.next();
+        if (header.values().length < 2) {
+            throw new OpenSearchException("geoip database should have at least two fields");
+        }
+        return header;
+    }
+
+    /***
+     * Update datasource as succeeded
+     *
+     * @param manifest the manifest
+     * @param datasource the datasource
+     * @return
+     * @throws IOException
+     */
+    private void updateDatasourceAsSucceeded(
+        final Datasource datasource,
+        final DatasourceManifest manifest,
+        final List<String> fields,
+        final Instant startTime,
+        final Instant endTime
+    ) throws IOException {
+        datasource.setDatabase(manifest, fields);
+        datasource.getUpdateStats().setLastSucceededAt(endTime);
+        datasource.getUpdateStats().setLastProcessingTimeInMillis(endTime.toEpochMilli() - startTime.toEpochMilli());
+        datasource.enable();
+        datasource.setState(DatasourceState.AVAILABLE);
+        datasourceFacade.updateDatasource(datasource);
+        log.info("GeoIP database creation succeeded for {} and took {} seconds", datasource.getId(), Duration.between(startTime, endTime));
+    }
+
+    /***
+     * Setup index to add a new geoip data
+     *
+     * @param manifest the manifest
+     * @param datasource the datasource
+     * @return
+     * @throws IOException
+     */
+    private String setupIndex(final DatasourceManifest manifest, final Datasource datasource) throws IOException {
+        String indexName = datasource.indexNameFor(manifest);
+        datasource.getIndices().add(indexName);
+        datasourceFacade.updateDatasource(datasource);
+        geoIpDataFacade.createIndexIfNotExists(indexName);
+        return indexName;
+    }
+
+    /**
+     * Determine if update is needed or not
+     *
+     * Update is needed when all following conditions are met
+     * 1. MD5 hash value in datasource is different with MD5 hash value in manifest
+     * 2. updatedAt value in datasource is before updateAt value in manifest
+     *
+     * @param datasource
```

Comment:
Your comments suggest that your method is shouldUpdate but the logic is skipUpdate. Can you keep it as consistency?

Reply:
Done